### ISSUE TITLE
Xnor trained models require different conversion tools?

### ISSUE THREAD

**Issue by prateethvnayak** [AlexeyAB/darknet#1688]
I was trying to train a custom model with Xnor cfg files. 
I need to deploy it into iOS platform. So do the Xnor trained models require some changes while converting to Keras model and then to CoreML model ? 

I was successfully able to get it to .mlmodel but it gives Build Errors in Xcode saying cannot read the model file. There are no errors with the Normal trained models. Only Xnor Trained models give this error in Xcode.

**Comment by AlexeyAB** [comment]
There should be implemented special Convolutional or GEMM algorithms to use XNOR-models, which are only in the Darknet.
Also accelerated Convolutional/GEMM algorithms are only in this repository.

**Comment by prateethvnayak** [comment]
@AlexeyAB 
Also, the trained xnor models should apparently give faster inference. But, the tiny_yolov3 xnor trained model that I have done (~20k iterations) on custom dataset with 4 classes is still slower than the normal tiny-yolo. 
Any comments on that ? 
I am using Nvidia Quadro-p6000 gpus. 

The above observation is using the `detector map` function.

**Comment by AlexeyAB** [comment]
* XNOR on GPU is only slightly faster currently, it will be improved. On CPU it is 3-4x faster.

* Do you use the latest version of this repository? And the latest version of https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3-tiny_xnor.cfg you can use it with previously trained model, there are added `bin_output=1` params in some layers.

* Can you show screenshots with these results, that `yolov3-tiny.cfg` is faster than `yolov3-tiny.cfg`?

**Comment by prateethvnayak** [comment]
Yes,  I just cloned the repo today and built it with GPU support. Yes, it has the bin_output flags too. I just changed the filters as required for the Custom Training.

Is there a diff way to perform `detector map` on CPU only when the repo is built with GPU flag ? 

Here are the two screenshots for the validation on the same validation set of images - Inference time is lesser for the normal model.

Below image is for xnor trained yolov3-tiny model for 20k iterations on two gpus

![image](https://user-images.githubusercontent.com/34733770/46263803-96b21280-c4d9-11e8-9aaa-1f1dadae650d.png)

Below is the image of normal model yolov3-tiny model map outcome

![image](https://user-images.githubusercontent.com/34733770/46263814-c52fed80-c4d9-11e8-9030-e269a03ac640.png)

Also, there is drop in mAP. Or I am doing the whole thing wrong. Recently came across [this repo](https://github.com/AlexeyAB/yolo2_light) of yours, Anything specifically different in this ?

**Comment by AlexeyAB** [comment]
> Below is the image of normal model yolov3-tiny model map outcome
> 
> ![image](https://user-images.githubusercontent.com/34733770/46263814-c52fed80-c4d9-11e8-9030-e269a03ac640.png)
> 

Can you attach image with Total Detection Time?

Current cfg-files and implementation are experemental. 
May be later I will accelerate it. Also some layers which will not be accelerated, I will leave them float (32-bit), so there will be higher mAP (some layers fp32) and higher speed (accelerated).


----

> Recently came across this repo of yours, Anything specifically different in this ?

Only that - in this repo https://github.com/AlexeyAB/yolo2_light is implemented INT8 inference with about +30% speedup and only -1% mAP accuracy drop.

**Comment by prateethvnayak** [comment]
@AlexeyAB 
Hey, You are right, I just ran both again with flag `-gpus 0,1` and I got 1s speedup. Here are the screenshots.

Below is xnor trained model:
![image](https://user-images.githubusercontent.com/34733770/46264102-2659c000-c4de-11e8-8ed0-a7774e2c1f69.png)

Below is normal trained tiny-yolov3 model:
![image](https://user-images.githubusercontent.com/34733770/46264106-31145500-c4de-11e8-928e-4b3195a90ae6.png)

> Current cfg-files and implementation are experemental.
May be later I will accelerate it. Also some layers which will not be accelerated, I will leave them float (32-bit), so there will be higher mAP (some layers fp32) and higher speed (accelerated).
Great, Thanks for the input on this.
I will have to look into conversion of xnor trained models to Keras and then CoreML model. Any good resources for these ? 
Thanks
So, the yolo2_light repo is only for INT8 Inference then ? can I try the config file and trained model from this repo with yolo2_light repo ?

**Comment by AlexeyAB** [comment]
@prateethvnayak 

> can I try the config file and trained model from this repo with yolo2_light repo ?

Yes, you can. 

* yolo2_light: only for Yolo, with Float-32, INT-8, XNOR-bit-1
* This repo: Float-32, TensorCores (FP16/32), XNOR-bit-1


> Any good resources for these ?

I don't know resources about using XNOR bit-1 inference on Keras/CoreML.
